diff --git a/feeds/packages/libs/gnutls/Makefile b/feeds/packages/libs/gnutls/Makefile
index 211470205..1634e9681 100644
--- a/libs/gnutls/Makefile
+++ b/libs/gnutls/Makefile
@@ -6,13 +6,13 @@
 include $(TOPDIR)/rules.mk

 PKG_NAME:=gnutls
-PKG_VERSION:=3.7.8
-PKG_RELEASE:=$(AUTORELEASE)
+PKG_VERSION:=3.8.0
+PKG_RELEASE:=1
 PKG_USE_MIPS16:=0

 PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION).tar.xz
-PKG_SOURCE_URL:=https://www.gnupg.org/ftp/gcrypt/gnutls/v3.7
-PKG_HASH:=c58ad39af0670efe6a8aee5e3a8b2331a1200418b64b7c51977fb396d4617114
+PKG_SOURCE_URL:=https://www.gnupg.org/ftp/gcrypt/gnutls/v3.8
+PKG_HASH:=0ea0d11a1660a1e63f960f157b197abe6d0c8cb3255be24e1fb3815930b9bdc5

 PKG_MAINTAINER:=Nikos Mavrogiannopoulos <nmav@gnutls.org>
 PKG_LICENSE:=LGPL-2.1-or-later
@@ -21,7 +21,7 @@ PKG_CPE_ID:=cpe:/a:gnu:gnutls

 PKG_BUILD_DEPENDS:=gettext-full/host
 PKG_BUILD_PARALLEL:=1
-PKG_FIXUP:=autoreconf gettext-version
+
 PKG_INSTALL:=1
 PKG_LIBTOOL_PATHS:=. lib

@@ -128,7 +128,6 @@ CONFIGURE_ARGS+= \
        --disable-valgrind-tests \
        --disable-ssl2-support \
        --disable-ssl3-support \
-       --enable-local-libopts \
        --without-idn \
        --with-default-trust-store-dir=/etc/ssl/certs/ \
        --with-included-unistring \
