--- a/target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-r619ac.dtsi	2022-05-10 20:53:55.083368146 +0800
+++ b/target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-r619ac.dtsi	2022-05-10 20:55:47.324080783 +0800
@@ -187,6 +187,9 @@
                                                reg = <0x5000 0x2f20>;
                                        };
                                };
+                        partition@180000 {
+                                label = "unused";
+                                reg = <0x180000 0xe80000>;
                        };
                };
        };
