--- natflow_user.c	2022-05-12 07:19:47.516405279 +0800
+++ natflow_user.c.new	2022-05-12 07:26:02.236808139 +0800
@@ -1916,7 +1916,7 @@
 			for (i = user->next_bucket; i < hashsz; i++) {
 				hlist_nulls_for_each_entry_rcu(h, n, &ct_hash[i], hnnode) {
 					ct = nf_ct_tuplehash_to_ctrack(h);
-					if (unlikely(!REFCOUNT_inc_not_zero(&ct->ct_general.use)))
+					if (unlikely(!atomic_inc_not_zero(&ct->ct_general.use)))
 						continue;
 
 					/* we only want to print DIR_ORIGINAL */
@@ -2076,7 +2076,7 @@
 		for (i = user->next_bucket; i < hashsz; i++) {
 			hlist_nulls_for_each_entry_rcu(h, n, &ct_hash[i], hnnode) {
 				ct = nf_ct_tuplehash_to_ctrack(h);
-				if (unlikely(!REFCOUNT_inc_not_zero(&ct->ct_general.use)))
+				if (unlikely(!atomic_inc_not_zero(&ct->ct_general.use)))
 					continue;
 
 				/* we only want to print DIR_ORIGINAL */
