--- a/src/tls.h	2018-03-07 14:35:05.000000000 +0800
+++ b/src/tls.h	2020-09-10 19:25:47.147381349 +0800
@@ -28,6 +28,6 @@
 
 #include "protocol.h"
 
-const protocol_t *const tls_protocol;
+extern const protocol_t *const tls_protocol;
 
 #endif

--- a/src/http.h	2018-03-07 14:35:05.000000000 +0800
+++ b/src/http.h	2020-09-10 19:24:46.129499253 +0800
@@ -29,6 +29,6 @@
 #include <stdio.h>
 #include "protocol.h"
 
-const protocol_t *const http_protocol;
+extern const protocol_t *const http_protocol;
 
 #endif
