--- a/os_dep/linux/ioctl_cfg80211.c
+++ b/os_dep/linux/ioctl_cfg80211.c
@@ -466,7 +466,7 @@
 
                cfg80211_ch_switch_started_notify(adapter->pnetdev, &chdef, 0, false);
 #else
-               cfg80211_ch_switch_started_notify(adapter->pnetdev, &chdef, 0);
+               cfg80211_ch_switch_started_notify(adapter->pnetdev, &chdef, 0, false);
 #endif
                goto exit;
        }
@@ -5017,7 +5017,7 @@
        mon_wdev->iftype = NL80211_IFTYPE_MONITOR;
        mon_ndev->ieee80211_ptr = mon_wdev;
 
-       ret = register_netdevice(mon_ndev);
+       cfg80211_register_netdevice(mon_ndev);
        if (ret)
                goto out;
 
@@ -6706,7 +6706,7 @@
                pwdev_priv = adapter_wdev_data(adapter);
 
                if (ndev == pwdev_priv->pmon_ndev) {
-                       unregister_netdevice(ndev);
+                       cfg80211_unregister_netdevice(ndev);
                        pwdev_priv->pmon_ndev = NULL;
                        pwdev_priv->ifname_mon[0] = '\0';
                        RTW_INFO(FUNC_NDEV_FMT" remove monitor ndev\n", FUNC_NDEV_ARG(ndev));
