--- a/Makefile  2018-12-04 01:58:09.000000000 +0800
+++ b/Makefile  2018-12-04 04:51:29.724865126 +0800
@@ -885,7 +885,7 @@
 	@rm -rf $(ROOTDIR)/data/dvb-scan/*
 	@$(ROOTDIR)/support/getmuxlist $(ROOTDIR)/data/dvb-scan
 	@touch $@
-
+	@mv $(ROOTDIR)/data/dvb-scan/dvb-c/mk-Telekabel-Sveti* $(ROOTDIR)/data/dvb-scan/dvb-c/mk-Telekabel-Sveti_Nikole        
 .PHONY: check_dvb_scan
 check_dvb_scan: $(ROOTDIR)/data/dvb-scan/.stamp
 
