--- a/include/drv_types.h
+++ b/include/drv_types.h
@@ -242,7 +242,7 @@
 #ifdef CONFIG_TX_EARLY_MODE
 	u8   early_mode;
 #endif
-#ifdef CONFIG_RTW_SW_LED
+#ifdef CONFIG_SW_LED
 	u8   led_ctrl;
 #endif
 #ifdef CONFIG_NARROWBAND_SUPPORTING
@@ -457,7 +457,7 @@
 #ifdef CONFIG_DFS_MASTER
 	u8 dfs_region_domain;
 #endif
-
+	u8 amsdu_mode;
 #ifdef CONFIG_MCC_MODE
 	u8 en_mcc;
 	u32 rtw_mcc_single_tx_cri;
@@ -1464,6 +1464,16 @@
 	u32 rcr_bf_suspend;
 	u32 cr_ext_bf_suspend;
 #endif /* CONFIG_WOWLAN */
+
+#if defined (CONFIG_CONCURRENT_MODE)  && defined (CONFIG_TSF_SYNC)
+	u16 sync_tsfr_counter;
+#endif
+
+	/* WPAS maintain from w1.fi */
+#define RTW_WPAS_W1FI		0x00
+	/* WPAS maintain from android */
+#define RTW_WPAS_ANDROID	0x01
+	u8 wpas_type;
 };
 
 #define DEV_STA_NUM(_dvobj)			MSTATE_STA_NUM(&((_dvobj)->iface_state))
@@ -1920,7 +1930,7 @@
 #endif
 
 #define adapter_mac_addr(adapter) (adapter->mac_addr)
-#ifdef CONFIG_RTW_CFGVENDOR_RANDOM_MAC_OUI
+#if defined(CONFIG_RTW_CFGVENDOR_RANDOM_MAC_OUI) || defined(CONFIG_RTW_SCAN_RAND)
 #define adapter_pno_mac_addr(adapter) \
 	((adapter_wdev_data(adapter))->pno_mac_addr)
 #endif
