--- a/Makefile	2023-07-11 11:34:42.000000000 +0800
+++ b/Makefile	2024-07-17 19:10:58.136163728 +0800
@@ -19,7 +19,8 @@
 EXTRA_CFLAGS += -Wno-implicit-fallthrough
 #EXTRA_CFLAGS += -Wno-return-type
 #EXTRA_CFLAGS += -Wno-discarded-qualifiers
-
+EXTRA_CFLAGS += -Wno-missing-prototypes
+EXTRA_CFLAGS += -Wno-missing-declarations
 # Activates Concurrent Mode if uncommented
 #EXTRA_CFLAGS += -DCONFIG_CONCURRENT_MODE
 
@@ -32,6 +33,12 @@
 EXTRA_CFLAGS += -Wno-address
 EXTRA_CFLAGS += -Wframe-larger-than=1648
 
+# gcc-13
+# EXTRA_CFLAGS += -Wno-enum-int-mismatch
+# EXTRA_CFLAGS += -Wno-stringop-overread
+# EXTRA_CFLAGS += -Wno-enum-conversion
+# EXTRA_CFLAGS += -Wno-int-in-bool-context
+
 GCC_VER_49 := $(shell echo `$(CC) -dumpversion | cut -f1-2 -d.` \>= 4.9 | bc )
 ifeq ($(GCC_VER_49),1)
 EXTRA_CFLAGS += -Wno-date-time	# Fix compile error && warning on gcc 4.9 and later
